pegen/pegen.h

```diff
@@ -103,13 +102,21 @@ void *dedent_token(Parser *p);
 expr_ty number_token(Parser *p);
 void *string_token(Parser *p);
 int raise_syntax_error(Parser *p, const char *errmsg, ...);
-
 void *CONSTRUCTOR(Parser *p, ...);
 
 #define UNUSED(expr) do { (void)(expr); } while (0)
 #define EXTRA_EXPR(head, tail) head->lineno, head->col_offset, tail->end_lineno, tail->end_col_offset, p->arena
 #define EXTRA start_lineno, start_col_offset, end_lineno, end_col_offset, p->arena
 
+inline void *
+CHECK_CALL(Parser *p, void *result)
+{
+    if (result == NULL) {
+        longjmp(p->error_env, 1);
+    }
```

Comment:
```suggestion
    if (result == NULL) {
        assert(PyErr_Occurred());
        longjmp(p->error_env, 1);
    }
```

Reply:
Should this be an assertion or should it be in the if condition? This way, we could use `CHECK_CALL` for *every* nested call. Some functions are allowed and should be able to return NULL, like get_keys and get_values in `dictrule`.

Reply:
>  Some functions are allowed and should be able to return NULL, like get_keys and get_values in dictrule.

Those return always an `asdl_seq *` that is freshly created, no?

```
asdl_seq *
get_keys(Parser *p, asdl_seq *seq)
{
    int len = asdl_seq_LEN(seq);
    asdl_seq *new_seq = _Py_asdl_seq_new(len, p->arena);
    if (!new_seq) {
        return NULL;
    }
    for (int i = 0; i < len; i++) {
        KeyValuePair *pair = asdl_seq_GET(seq, i);
        asdl_seq_SET(new_seq, i, pair->key);
    }
    return new_seq;
}
```

What can be NULL there?

It should be an assertion because we don't need to differentiate between `NULL`-as-error and some other meaning of `NULL`. If a helper returns `NULL` is always an error. The assert will allow us to detect cases when we are not setting the error indicator or something like that.

Reply:
That's right. I was getting some errors in `simpy_cpython` that were related to the empty `dict`. I had initially included the `CHECK_CALL` call for `get_keys` and `get_values` and the errors went away, when I removed those. At the time, I assumed that `_Py_asdl_seq_new` returns NULL, if called with `size=0`, but that's not the case.

I now added them back and everything runs smoothly, so it must have been something else, but I don't know exactly what.

In any case, my bad.